integration/kotlinx-coroutines-play-services/src/Tasks.kt

```diff
@@ -91,17 +136,28 @@ public suspend fun <T> Task<T>.await(): T {
         } else {
             throw e
         }
+    } else if (cancellationTokenSource.token.isCancellationRequested) {
+        // The task hasn't completed, yet cancellation was already requested.
+        // Interpret this by throwing immediately (no way to cancel the task)
+        throw CancellationException("Cancellation was already requested")
     }
 
     return suspendCancellableCoroutine { cont ->
         addOnCompleteListener {
-            val e = exception
+            val e = it.exception
             if (e == null) {
                 @Suppress("UNCHECKED_CAST")
-                if (isCanceled) cont.cancel() else cont.resume(result as T)
+                if (it.isCanceled) cont.cancel() else cont.resume(it.result as T)
             } else {
                 cont.resumeWithException(e)
             }
         }
+        cancellationTokenSource.token.onCanceledRequested {
+            cont.cancel()
```

Comment:
Same reasoning here: we should wait until task is completed, so we are sure it hasn't thrown an exception after cancellation